doc/decisions/0_drafts/errors_core.md

```diff
@@ -0,0 +1,403 @@
+  Often this case can be excluded by static analysis, i.e., because of the preceding code it is impossible that the pointer is `NULL`.
+
+  ```c
+  int foo (KeySet * ks, Key * key) {
+    if (key == NULL) {
+      return -1;
+    }
+
+    // [...] could be lots of code, but the code doesn't change `Key * key`
+
+    const Key * found1 = ksLookupByName (ks, "/foo", 0);
+    if (found1 != NULL) {
+      // no need to worry about key == NULL, because we checked earlier
+      // also no need to worry about ks == NULL, because the ksLookupByName would have returned NULL
+      const Key * found2 = ksLookup (ks, key, 0);
+    }
+  }
+  ```
+
+- It is not always necessary to report separate error codes, for _every_ error case.
+  In many cases, the caller can distinguish after the fact, e.g., `keySetName` fails when the given name is invalid and when the given `Key` has a read-only name.
+  This can be distinguished after the fact, by checking whether the `Key` has a read-only name with `keyIsLocked`.
+
+- Callers normally don't care about the exact error, unless they can do something about it.
+  This could be:
+
+  - reporting the error to a user
+  - switching to a different code path
+  - (during debugging) changing the code
+
+- Functions can be written in a such a way that they detect and report all errors without permanent effects.
+
+- The functions that cannot use a `Key * errorKey`, are side-effect-free.
+  Therefore, the only possible cause of errors are invalid arguments.
+
+## Considered Alternatives
+
+### Extra argument
+
+All functions that report errors take a `int * error` argument to which errors can be written.
+
+This makes the API much more convoluted to use.
+
+A benefit would be the return value remains fully usable.
+
+### Thread Local Storage
+
+Use a field in thread local storage (TLS) to store errors similar to `errno`.
+
+This leaves the API design entirely open.
+
+TLS was only standardized in C11, while Elektra currently only requires C99.
+
+This has all the same issues as `errno`: needs to be cleared to detect errors, can only store one error (per thread), etc.
+
+### Limit to single error value
+
+Always limiting a function to a single error code can be very limiting and in some cases and makes it very hard to find out what the exact problem was.
+Reducing a function to a single error value, makes sense if the errors are easy to distinguish.
+But for more complex preconditions, this can be annoying and inefficient to use.
+For example, a function `foo` that takes two keynames, both of which must be valid.
+To find out which of the names was invalid, the caller would have to call something like `elektraCheckName(name)` on at least one name.
+That's a non-trivial check that is not cheap in terms of runtime.
+More importantly, it is a check that `foo` already did, and the result could just be propagated to the caller.
+
+### Redesign API
```

Comment:
Such redesign is better done in #4805. The decision here simply would be "Limit to single error value" then.

Reply:
Not sure what you want to say here. I can rename the solution, if you like, but not sure what this is going to change...

```suggestion
### Limit to single error value
```